marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs

```diff
@@ -0,0 +1,311 @@
+/**
+ Copyright (c) 2020 MarkLogic Corporation
+
+ Licensed under the Apache License, Version 2.0 (the 'License');
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an 'AS IS' BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+'use strict';
+
+// Global constants
+const textStartsWith = '#';
+const textPropName = textStartsWith + 'text';
+const attrStartsWith = '@';
+const outputAttrStartsWith = '@'; // in case this ever needs to vary from the input.
+const nsPrefixDelim = ':';
+const nsAttrStartsWith = attrStartsWith + 'xmlns';
+const defaultNSAttrName = nsAttrStartsWith;
+const nonDefaultNSAttrStartsWith = nsAttrStartsWith + nsPrefixDelim
+
+// Set up for parser
+const parser = require('/fast-xml-parser/src/parser.js');
```

Comment:
Is all of this code coming from our UI? There's a ton of logic in here and I'm wondering how much automated test coverage we have of it.

Reply:
I wrote all of this code for this ticket.  I wasn't aware of existing code :/  Can you provide a pointer to that code?

Reply:
The UI has some of the code (as the requirements for endpoint is slightly different) in mapping-card.tsx and source-to-entity-map.tsx

Reply:
That's fine that you wrote it, I was just curious if it came from the UI first. 

The challenge here is there's a lot of code / conditionals / logic here, and I'm wondering how to ensure it's all covered correctly. There's one ml-unit-test for it, and while the XML input document has a lot of complexity to it, there's no way to know what the total set of scenarios is that we need to support. 

The tests can identify that though. Instead of having one large test that involves many scenarios, I think a better approach is to have a lot of small tests (they could all be separate functions in one file) that each test a single scenario. That would better identify what the total set of known scenarios that we support is.

Reply:
Thanks, gents.  I'll check out the existing code (as it is proven) and will work over the tests as suggested.

Reply:
Initial assessment of referenced code...

### marklogic-data-hub-central/ui/src/components/entities/mapping/mapping-card.tsx:

* When source format is XML, fast-xml-parser's output is passed into generateNestedDataSource()
* When source format is JSON, that JSON is passed into generateNestedDataSource() as is.
* generateNestedDataSource() traverses the given JSON and creates a new structure that's quite a bit different than 6098's spec.
* Seems like only the traversal logic could be compared with DHFPROD-6098's implementation.

### /marklogic-data-hub-central/ui/src/components/entities/mapping/source-entity-map/source-to-entity-map.tsx

* Receives sourceData (output from generateNestedDataSource()) and fetchSrcDocFromUri (function) from mapping-card.tsx.
* fetchSrcDocFromUri appears only used when changing source document (via navigation).
* sourceData given to flattenSourceDoc()
* flattenSourceDoc() is the UI's implementation of DHFPROD-6101.

Reply:
DHFPROD-6101 covers entirety of flattenSourceDoc() and some more. We will have to handle scenarios where properties are not Qnames in DHFPROD-6101.